### DISCUSSION TITLE
id64 string for Reality Models

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#3351] [Q&A] [answered]
Greetings, I have a realityDataId that I want to use in PlanerClipMaskSettings.create( ... ). But the problem is that the maskModel takes in id64 string and not the realityDataId. Is there any way to retrieve RealityDataId in id64 format?

**Comment by pmconne** [comment, marked as answer]
Please be specific about:
- What version of iTwin.js you are using.
- What specific API you are trying to use (I find no PlanarClipMaskSettings APIs in 2.x or 3.x that take an argument named `maskModel`).
- Most importantly, what you are trying to achieve (e.g., are you trying to apply a mask to the reality model, or mask some other model / background map using the reality model?).

It helps if you show your code. For example, if you are calling `DisplayStyleState.attachRealityModel`, that function returns a ContextRealityModelState which has a transient model Id that can be used in many APIs.
A RealityDataId is not a 64-bit integer and cannot be used in place of Id64String.

**Reply by hanan-commits-work** [reply]
Hi. I am using 3.x and following this [sandbox](https://www.itwinjs.org/sandboxes/pmconne/Inverted%20model%20mask). And i want to mask the model using reality model.

I want to use the id of the currently active reality model to mask the other elements.

**Reply by pmconne** [reply]
Using a reality model to mask out (hide) portions of a design model is not currently supported, per [the docs](https://www.itwinjs.org/reference/core-common/displaystyles/planarclipmasksettings/):
```
 * @note Currently reality models (including background maps and terrain) can be masked, but design models can only produce masks.
```

**Reply by hanan-commits-work** [reply]
let me clarify my statement...
I want to use reality models as the source, and the CAD models as the mask. so i am trying to apply the mask to the reality model.
sorry for the ambiguous statement earlier.

**Reply by ovaisa** [reply]
Hi Paul, bit more clarification.  Referring to third point in your response above

- Most importantly, what you are trying to achieve (e.g., are you trying to apply a mask to the reality model, or mask some other model / background map using the reality model?).

We are trying to apply a mask to the reality model.

We have a reality model of a plant with equipment and piping.   Our use case is that we want to remove the equipment from the reality model.  We have a CAD element that encapsulates this equipment and we are trying to use it as mask to remove the equipment from the model.

Ovais

**Reply by pmconne** [reply]
If you edit the sandbox you linked as follows:
```
    const mask = PlanarClipMaskSettings.create({
      //invert:  true,
      modelIds: [maskModel],
      //elementIds: maskElem,
      transparency: 0.5,
    });
```
You will observe the background map is semi-transparent wherever it intersects the specified design model.
Note this is achieved by setting [BackgroundMapProps.planarClipMask](https://www.itwinjs.org/reference/core-common/displaystyles/backgroundmapprops/planarclipmask/) when calling changeBackgroundMapProps.
The same thing can be achieved with a reality model by setting [ContextRealityModelProps.planarClipMask](https://www.itwinjs.org/reference/core-common/displaystyles/contextrealitymodelprops/planarclipmask/) when calling [DisplayStyleState.attachRealityModel](https://www.itwinjs.org/reference/core-frontend/views/displaystylestate/attachrealitymodel/).

As noted, you cannot turn the other **design** models transparent using a planar clip mask. You can do so by using symbology overrides, e.g. via [EmphasizeElements](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=emphasize-elements-sample&imodel=Retail+Building+Sample).

**Reply by hanan-commits-work** [reply]
Hi. 
I tried to mimic your advise. But my application keeps on crashing after setting the mask. We are getting an assertion error after clicking on the CAD model, in order to create the mask. So I created a sample app to understand what exactly maybe the problem. But even the sample app crashed.
I am attaching the code.
[sample_app.zip](https://github.com/iTwin/itwinjs-core/files/8251128/sample_app.zip)

Let me walk you through the code.
1. When you run the app, you would see a green button in the widgets to the right.
2. Click on the button and then click anywhere on the grid. 
3. It will open a window asking you to place a glb asset on the grid. 
4. Go the public folder of the sample_app and select model.glb
5. A 3d object will be placed.
6. Now the mask is applied after we click on the 3d object.
7. Click on the object and youd see everything other than the selected object turn to blue. 
8. This is where the application crashes.


**As for the code, youd need to look into two files only**
1. View.tsx
2. gLTFDecorator.tsx

**View.tsx:**
  Kindly find the function makeMask. It accepts the PlanarClipMaskSettings object and then creates the mask.

**gLTFDecorator.tsx:**
   Kindly find the function createMask, it is called by onDecorationButtonEvent function.


Let me know if you have further questions regarding the code.

**Reply by pmconne** [reply]
> We are getting an assertion error

Provide the error.

**Reply by hanan-commits-work** [reply]
![image](https://user-images.githubusercontent.com/89520756/158412689-4b50da1a-ad3d-4fbc-8263-e1eba66fb904.png)
This is the error.

**Reply by pmconne** [reply]
Are you certain your model Id identifies a geometric model stored in the iModel?
```
  public getTileTrees(view: ViewState3d, classifiedModelId: Id64String): TileTreeReference[] | undefined {
    if (this.settings.mode === PlanarClipMaskMode.Priority) {
      const viewTrees = new Array<TileTreeReference>();
      const thisPriority = this.settings.priority === undefined ? PlanarClipMaskPriority.RealityModel : this.settings.priority;
      view.forEachTileTreeRef((ref) => {
        const tree = ref.treeOwner.load();
        if (tree && tree.modelId !== classifiedModelId && ref.planarclipMaskPriority > thisPriority)
          viewTrees.push(ref);
      });

      return viewTrees;
    }

    if (!this._tileTreeRefs) {
      this._tileTreeRefs = new Array<TileTreeReference>();
      if (this.settings.modelIds) {
        for (const modelId of this.settings.modelIds) {
          const model = view.iModel.models.getLoaded(modelId);
          assert(model !== undefined);   // Models should be loaded by RealityModelTileTree
```

**Reply by hanan-commits-work** [reply]
let me confirm it again.
